[Gutenberg] Add Sentry React native SDK

fluiddot opened the pull request:
`gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3629
`gutenberg` PR: https://github.com/WordPress/gutenberg/pull/32768
`Automattic-Tracks-iOS` PR: https://github.com/Automattic/Automattic-Tracks-iOS/pull/183

To test:
Follow the testing instructions from `gutenberg` PR: https://github.com/WordPress/gutenberg/pull/32768

## Regression Notes
1. Potential unintended areas of impact
The lanes (`Fastfile`) related to building and uploading the build as they're using a new lane I've added for uploading the source maps to Sentry.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I manually tested the lane by running the command `bundle exec fastlane upload_gutenberg_sourcemaps internal:true` but I haven't tested the ones related to building and uploading builds.

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

fluiddot committed:
Add bridge methods for React native crash logging

fluiddot committed:
Autocorrect lint issues

fluiddot committed:
Add getSentryUser to Gutenberg view controller

fluiddot committed:
Add upload Gutenberg source maps lane to Fastfile

fluiddot committed:
Update Tracks ref

fluiddot commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -930,6 +930,12 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
     }
+
+    func gutenbergDidLogSentryEnvelope(_ envelope: [String: Any]) {
+        DispatchQueue.main.async {
+            WordPressAppDelegate.crashLogging?.logEnvelope(envelope)
+        }
```
Calls to `WordPressAppDelegate` are required to be done in the main thread.

fluiddot commented on fastlane/Fastfile:
```diff
@@ -664,6 +673,41 @@ import "./ScreenshotFastfile"
+      project_slug: 'wordpress-ios',
+      version: build_version,
+      dist: build_version,
+      rewrite: true,
+      strip_common_prefix: true,
```
When the React native bundle is generated, the source map file references include the local machine path, with these options it automatically strips this part.

mokagio replied:
Thank you for the info. I think it would be valuable to add them in the code, too:

```suggestion
      # When the React native bundle is generated, the source map file references
      # include the local machine path, with the `rewrite` and `strip_common_prefix` 
      # options Sentry automatically strips this part.
      rewrite: true,
      strip_common_prefix: true,      
```

fluiddot commented on fastlane/Fastfile:
```diff
@@ -664,6 +673,41 @@ import "./ScreenshotFastfile"
+    sourcemaps_folder = File.join(gutenberg_bundle, "sourcemaps")
+    FileUtils.mkdir_p(sourcemaps_folder)
+    # Name of bundle file must be set to main.jsbundle
+    FileUtils.cp(File.join(gutenberg_bundle, 'App.js'), File.join(sourcemaps_folder, 'main.jsbundle'))
+    FileUtils.cp(File.join(gutenberg_bundle, 'App.js.map'), File.join(sourcemaps_folder, 'main.jsbundle.map'))
```
It's important that the source map files have specific names, otherwise, Sentry doesn't un-minify the stack traces.

fluiddot commented on fastlane/Fastfile:
```diff
@@ -664,6 +673,41 @@ import "./ScreenshotFastfile"
+  # bundle exec fastlane upload_gutenberg_sourcemaps internal:true
+  #####################################################################################
+  lane :upload_gutenberg_sourcemaps do | options |
+    build_version = ios_get_build_version(internal: options[:internal])
```
The build version is different depending on the `internal` option, here is an example:
- `internal: true`: `17.6.0.20210614` ([reference](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/config/Version.internal.xcconfig))
- `internal: false`: `17.6.0.0` ([reference](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/config/Version.public.xcconfig))

fluiddot committed:
Update Podfile.lock

fluiddot committed:
Update Gutenberg Mobile ref

fluiddot committed:
Change internal value in build_and_upload_internal lane

fluiddot committed:
Merge branch 'develop' into gutenberg/add/sentry-react-native

# Conflicts:
#	Podfile
#	Podfile.lock
#	fastlane/Fastfile

fluiddot committed:
Update Gutenberg Mobile ref

fluiddot commented:
@mokagio @jkmassel I've added a new lane to the `Fastfile` to upload the React native source maps to Sentry, I'd appreciate it if you could take a look at that part just in case I'm overlooking anything, thanks 🙇 !

antonis reviewed (approved):
Tested via https://github.com/WordPress/gutenberg/pull/32768#pullrequestreview-692939310 
The code changes LGTM too 🎉

mokagio commented on fastlane/Fastfile:
```diff
@@ -418,6 +418,9 @@ platform :ios do
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
 
+    # Upload Gutenberg source maps
+    upload_gutenberg_sourcemaps(internal: true)
```
Nitpick: given the well chosen method name, the comment is redundant 😄 

```suggestion
    upload_gutenberg_sourcemaps(internal: true)
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -474,6 +477,9 @@ platform :ios do
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
+
+    # Upload Gutenberg source maps
+    upload_gutenberg_sourcemaps(internal: true)
```
Nitpick: given the well chosen method name, the comment is redundant 😄 

```suggestion
    upload_gutenberg_sourcemaps(internal: true)
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -519,6 +525,9 @@ platform :ios do
       dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
     )
 
+    # Upload Gutenberg source maps
+    upload_gutenberg_sourcemaps(internal: false)
```
Nitpick: given the well chosen method name, the comment is redundant 😄 

```suggestion
    upload_gutenberg_sourcemaps(internal: false)
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -677,6 +686,41 @@ platform :ios do
+    gutenberg_bundle = File.join(PROJECT_ROOT_FOLDER, "Pods/Gutenberg/bundle/ios")
+    sourcemaps_folder = File.join(gutenberg_bundle, "sourcemaps")
+    FileUtils.mkdir_p(sourcemaps_folder)
+    # Name of bundle file must be set to main.jsbundle
+    FileUtils.cp(File.join(gutenberg_bundle, 'App.js'), File.join(sourcemaps_folder, 'main.jsbundle'))
+    FileUtils.cp(File.join(gutenberg_bundle, 'App.js.map'), File.join(sourcemaps_folder, 'main.jsbundle.map'))
```
Thanks for the info. I think this comment would be useful in the lane, too.

```suggestion
    # Name of bundle file must be set to main.jsbundle.
    #
    # It's important that the source map files have specific names, otherwise, Sentry 
    # won't un-minify the stack traces.
    FileUtils.cp(File.join(gutenberg_bundle, 'App.js'), File.join(sourcemaps_folder, 'main.jsbundle'))
    FileUtils.cp(File.join(gutenberg_bundle, 'App.js.map'), File.join(sourcemaps_folder, 'main.jsbundle.map'))
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -677,6 +686,41 @@ platform :ios do
+      sourcemap: sourcemaps_folder
+    )
+
+    FileUtils.rm_rf(sourcemaps_folder)
```
Ruby suggestion: there's a way to run this code that depends on a temporary directory in a block, so that even if something fails, the runtime will ensure the directory is removed.

It should be something like:

```ruby
require 'tmpdir'

Dir.mktmpdir do |sourcemaps_folder|
  # all the code should "just work" the way it's written right now
end
```

fluiddot replied:
Thanks for the suggestion 🙇 ! I'll update the code to use the temporary directory block.

fluiddot replied:
I applied your suggestion in [this commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/16700/commits/20f1892b50477fb1a415e0991e7135c5e99a1aaf) and run some tests, it worked like charm 🎊 .

mokagio commented on fastlane/Fastfile:
```diff
@@ -677,6 +686,41 @@ platform :ios do
+      auth_token: get_required_env("SENTRY_AUTH_TOKEN"),
+      org_slug: 'a8c',
+      project_slug: 'wordpress-ios',
+      version: build_version,
+      dist: build_version,
```
Just to check, what's the difference between these two parameters and why is it okay for them to have the same value?

fluiddot replied:
They are documented [here](https://docs.sentry.io/platforms/react-native/manual-setup/sourcemaps/#configure-cli), the `version` value should match the one set [here](https://github.com/Automattic/Automattic-Tracks-iOS/blob/aa7040e504377a0322dd91ff366f115e9873a369/Automattic-Tracks-iOS/Crash%20Logging/CrashLogging.swift#L39). However, I'm not sure about what to use for `dist`, when I checked Sentry events, I found that they had the same value (see attached examples) so I went ahead and use the same version value.

<img width="1003" alt="Screenshot 2021-07-01 at 12 48 22" src="https://user-images.githubusercontent.com/14905380/124114781-4642b780-da6d-11eb-8c01-0d8259f7d4cf.png">

I've just checked from where the Sentry iOS SDK takes the `dist` value ([code reference](https://github.com/getsentry/sentry-cocoa/blob/b750f8305717eeb46a11de89b15e2adddc8acc8e/Sources/Sentry/SentryClient.m#L367)) and it's `CFBundleVersion`, so looks like that in the end it's actually the same value 😄 . One thing I could try is to remove the `dist` parameter from the action and check if it affects somehow 👍.

fluiddot replied:
I've done a test by removing the `dist` parameter and so far it works, however, this change also implies initializing the SDK in the RN side without that parameter ([code reference](https://github.com/wordpress-mobile/gutenberg-mobile/blob/ca6d40ffcd02a18bec7aab258d76c4147ff6e2a3/src/sentry/index.js#L36)).

To be honest, I don't have a strong opinion on whether include it or not, wdyt?

mokagio replied:
@fluiddot I dug a bit into the documents and found a definition of what the `dist` value is meant to be [here](https://docs.sentry.io/platforms/react-native/configuration/releases/) which matches your finding:

> You will also need to pass a distribution to `dist` for events to be attributed correctly to a release. Usually this value is the build number, for example 52. This value needs to be unique to only each release.

Given that, your original approach of passing the build number was spot on.

The snippet in that pages shows the release as something that looks more like a node module name that the build number

```js
Sentry.init({
  release: "my-project-name@2.3.12",
  dist: "52",
});
```

I guess they're making a bit of confusion on their end between `release`, `dist`, `version`, and `build` in the attempt of supporting different platforms? 🤷‍♂️ 

From what I see in your explanation and screenshots, this setup seems fine. 👍

mokagio reviewed (commented):
Thank you for jumping into adding logic to the Fastlane automation @fluiddot 👏 🙇‍♂️ 

I only left nitpick / minor improvement comments

fluiddot committed:
Remove redundant comment from Fastfile file

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

fluiddot committed:
Remove redundant comment from Fastfile file

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

fluiddot committed:
Remove redundant comment from Fastfile file

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

fluiddot committed:
Add details to RN bundle naming in Fastfile file

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

fluiddot committed:
Add info comments to upload_gutenberg_sourcemaps lane

Include details about the use of `rewrite` and `strip_common_prefix` params of `sentry_upload_sourcemap` Fastlane action.

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

fluiddot committed:
Use temporary dir for uploading JS source maps

fluiddot committed:
Add Gutenberg Sentry native module

fluiddot committed:
Rename RNSentry to GutenbergSentry

fluiddot committed:
Update Gutenberg Mobile ref

fluiddot committed:
Apply swift lint autocorrection

fluiddot commented:
As posted in [this comment](https://github.com/WordPress/gutenberg/pull/32768#issuecomment-874051619), the Sentry native module has been moved to the WordPress-iOS project.

fluiddot committed:
Rename gutenberg sentry options function

fluiddot committed:
Update Gutenberg Mobile ref

fluiddot committed:
Merge branch 'develop' into gutenberg/add/sentry-react-native

# Conflicts:
#	Podfile
#	Podfile.lock

mokagio commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -265,8 +265,12 @@ class GutenbergViewController: UIViewController, PostEditor {
         return GutenbergImageLoader(post: post)
     }()
 
+    private lazy var gutenbergSentry: GutenbergSentry = {
+        return GutenbergSentry()
+    }()
```
This condensed syntax should work, too

```suggestion
    private lazy var gutenbergSentry = GutenbergSentry()
```

The result is the same, but I though it worth sharing because it's fun to see how the compiler unpacks the condensed syntax I suggested into the full one you used 😄

fluiddot committed:
Update WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

twstokes commented:
👋 I'm going to close this PR due to lack of activity, but please reopen if this was an error. Thank you.
